zcash_client_backend/src/decrypt.rs

```diff
@@ -49,27 +61,37 @@ pub fn decrypt_transaction<P: consensus::Parameters>(
     if let Some(bundle) = tx.sapling_bundle() {
         for (account, ufvk) in ufvks.iter() {
             if let Some(dfvk) = ufvk.sapling() {
-                let ivk = PreparedIncomingViewingKey::new(&dfvk.fvk().vk.ivk());
+                let ivk_external = PreparedIncomingViewingKey::new(&dfvk.to_ivk(Scope::External));
+                let ivk_internal = PreparedIncomingViewingKey::new(&dfvk.to_ivk(Scope::Internal));
                 let ovk = dfvk.fvk().ovk;
 
                 for (index, output) in bundle.shielded_outputs.iter().enumerate() {
-                    let ((note, to, memo), outgoing) =
-                        match try_sapling_note_decryption(params, height, &ivk, output) {
-                            Some(ret) => (ret, false),
-                            None => match try_sapling_output_recovery(params, height, &ovk, output)
-                            {
-                                Some(ret) => (ret, true),
-                                None => continue,
-                            },
-                        };
+                    let decryption_result =
+                        try_sapling_note_decryption(params, height, &ivk_external, output)
+                            .map(|ret| (ret, TransferType::Incoming))
+                            .or_else(|| {
+                                try_sapling_note_decryption(params, height, &ivk_internal, output)
+                                    .map(|ret| (ret, TransferType::WalletInternal))
+                            })
+                            .or_else(|| {
+                                try_sapling_output_recovery(params, height, &ovk, output)
+                                    .map(|ret| (ret, TransferType::Outgoing))
+                            });
```

Comment:
This looks vaguely like some similar code in `zcashd`. We should document the logic here, and try to keep it in line with what we do in `zcashd`.